### ISSUE TITLE
The memory used by the browser goes very high 5GB and 55 % CPU usage.

### ISSUE THREAD

**Issue by asharmaarcadix** [cypress-io/cypress#26096]
### Test code to reproduce

Attaching the video of chrome running where the test is timed-out and elements are not found in a single stretched script.

we are trying to run a Cypress script (written in cucumber) in one full stretch but it becomes too slow and leads to failure of the script due to group time end. 
when we try the same thing with multiple scenarios it works fast and fine.
But we want to have a single stretched scenario.
Can you please help us in this situation.

https://user-images.githubusercontent.com/121948808/224963112-d90b35c9-053c-41b7-a791-81884af3fdfb.mp4

[job-check-single.feature.txt](https://github.com/cypress-io/cypress/files/10966932/job-check-single.feature.txt)

The feature file is converted to txt file(For uploading purpose).

### Cypress Mode

both modes

### Cypress Version

12.2.0

### Browser Version

110.0.5481.180 

### Node version

v18.13.0

### Operating System

Windows 11 Pro

**Comment by lmiller1990** [comment]
How big exactly is your spec file? If I'm reading this correctly, the scenario takes 30+ minutes to run?

We have an experiment called `experimentalMemoryManagement` you could try. Docs: https://docs.cypress.io/guides/references/experiments#Configuration

In `cypress.config.ts`:

```js
module.exports = {
  e2e: {
    experimentalMemoryManagement: true
  }
}
```

**Comment by asharmaarcadix** [comment]
Hi, as mentioned, We tried using it.
But We are not getting a definition for keyword experimentalMemoryManagement.


Cypress Version details:-
![image](https://user-images.githubusercontent.com/121948808/225603325-c958ae8f-13ca-4737-8fda-7e999ff44ea6.png)

Image from IDE:-
![image](https://user-images.githubusercontent.com/121948808/225603679-7f6b6e83-f764-4697-9518-815d3752d333.png)

**Comment by lmiller1990** [comment]
It looks like you need to update to Cypress 12.4 to enable that experiment:

![image](https://user-images.githubusercontent.com/19196536/225807696-08f96d92-3e29-4d80-9f7d-44b053df0391.png)

**Comment by asharmaarcadix** [comment]
Hi, 
Thank you for solution.
I have Updated the Cypress version to latest, and i am not getting error anymore.
However, using the mentioned config in above, and running the test case, it gets slower as time goes on. (as Earlier)
Due to which my test case is getting failed.
Can you please help us with this.

**Comment by lmiller1990** [comment]
> I have Updated the Cypress version to latest, and i am not getting error anymore.

Great!

Can you clarify my earlier question: how big exactly is your spec file? If I'm reading this correctly, the scenario takes 30+ minutes to run? Is this a single test (eg, a single `it`) or a single `describe` with lots of tests (`it`) blocks?

I can't help out much more without knowing more about your project and code base - all I can see is a short video and screenshot. Please share some more information, starting with those questions above.

**Comment by asharmaarcadix** [comment]
[job-check-single.feature.txt](https://github.com/cypress-io/cypress/files/11019749/job-check-single.feature.txt)
Hi, please have a look at the feature file.( i have added txt extension just for Uploading purpose.)

It's a single test scenario. which we are trying to cover up.

**Comment by lmiller1990** [comment]
It looks like a fairly long and complex scenario. Each block starts with "When I click "next task" button". Is it possible to split these into separate tests/files, or is all must be in one `it` block?

This sounds similar to https://github.com/cypress-io/cypress/issues/2366. Historically the fix for huge memory usage has been [a configure option](https://docs.cypress.io/guides/references/configuration#Global), `numTestsKeptInMemory`, which you can set to `numTestsKeptInMemory: 0`. What this does is prevent Cypress from keeping too many snapshots in memory (each `cy.()` command takes a snapshot of the entire DOM, which takes a lot of memory over the course of many tests).

How long should it take for a user to complete the entire feature your test is expressing? 34 minutes for a single scenario feels very long? Also, does it complete when you run headlessly (using `npx cypress run`?). "Run" mode does NOT capture snapshots, which should consume far less memory. Can you let me know if it is able to complete in run mode?

**Comment by asharmaarcadix** [comment]
we want it to be a single block.
We tried with below config as menitoned.
![image](https://user-images.githubusercontent.com/121948808/226640860-dc916384-7657-4583-937b-6d4d96102956.png)

And Tried both "open" mode and "run" mode.
With Open mode we still face the same issue.
With Run mode we were not able to complete the test,
here is a screenshot from terminal.
![image](https://user-images.githubusercontent.com/121948808/226640734-b3279e66-0104-4f4f-9241-014e245d9974.png)

**Comment by lmiller1990** [comment]
I did some research. This is a bit of a unique case - generally we recommend shorter tests, since they lend themselves to a quicker feedback loop, as opposed to one long running test. I'll try to make some other suggestions, either way:

Here's some notes I found about Cucumber in Cypress:

As a note, there have been reports of [some users experiencing slowed performance](https://github.com/badeball/cypress-cucumber-preprocessor/issues/336) with Cucumber.
In most cases, there has been a direct correlation between slowed performance with Cucumber and a lack of machine resources, but I’d like to provide a few workarounds:
- Reduce numTestsKeptInMemory to 0 (or [lower than the default 50](https://docs.cypress.io/guides/references/configuration#Global)).
- Video options in your cypress.config
- Setting video: false in your cypress.config is the easiest way to free up resources.
  - If your team would still like to see the video conditionally, do NOT add video: false in your cypress.config. Instead, try the options below:
[Delete when specs pass](https://docs.cypress.io/api/plugins/after-spec-api#Delete-the-recorded-video-if-the-spec-passed). This will only upload videos for failed specs.
- [Delete when specs pass and have no flaky tests](https://docs.cypress.io/api/plugins/after-spec-api#Delete-the-recorded-video-if-no-tests-retried). This will upload both failed specs and videos for specs with flaky tests. (I personally like to use this one)
- Set videoCompression to 0 (or lower than the default setting of 32). You can find an example [here](https://docs.cypress.io/guides/guides/screenshots-and-videos#Video-encoding)
- If using a Chromium-based browser, set experimentalMemoryManagement to true

I think you've tried most of these already. When you execute using run mode, can you try turning OFF video to see if the test can at least pass? You could run it like this: `npx cypress run --browser chrome --config video=false`. 

Are you having issues in tests that do not use the cucumber plugin?  If it's an option, I'd recommend exploring writing some tests without it, and see if the problem continues. I'd like to see if this is an issue in Cypress, or only when using it in combination with the Cucumber plugin.

**Comment by lmiller1990** [comment]
Another thing to try, you can running like this:

```sh
CYPRESS_NO_COMMAND_LOG=1 npx cypress {run or open}
```

It will hide the log on the left side - the development experience won't be great, but it might help. I would highly recommend exploring some alternative options, a single test taking 30 minutes is going to be a struggle, even with all the optimizations and tricks available.

**Comment by lmiller1990** [comment]
Happy to reopen once we've got more information to continue debugging. Until then, I'll close this, since I don't see any actionable work in Cypress relating to this issue.

**Comment by LukasLewandowski** [comment]
We are still facing this problem when running even single test in headed mode. Cypress 13.5.0 with TypeScipt is able to eat 8 GB of memory with the experimental flag for memory management turned on and numTestsKeeptInMemory set only to 1. 
Both with node 20.9.0 and 18.17.1.
We are on OSX.
Our tests time execution are taking up to 9 minutes. 

I wonder why Cypress is not releasing memory between runs? I have noticed that I can execute one test which will pass and Cypress will grab for example 3 GB of my ram, then I will execute second test or return to list of tests and rerun the same test but memory usage will start from the same level, so 3 GB is already reserved for Cypress.

**Comment by jennifer-shehane** [comment]
@LukasLewandowski We will investigate if we can be given some information to recreate. A reproducible example, a link to a Cloud Test Replay. We're looking into performance atm.

**Comment by rakeshnambiar** [comment]
I can confirm that I am seeing the same issue on cypress `13.6.1` which is the current latest version during the local run on OSX and I have 16GB RAM.

I have the below settings in the config file.

```
  experimentalModifyObstructiveThirdPartyCode: true, 
  experimentalMemoryManagement: true,
  numTestsKeptInMemory: 0,
  ```
  when the tests starts it is printing:
  
  ```
    ────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        13.6.1                                                                         │
  │ Browser:        Electron 114 (headless)                                                        │
  │ Node Version:   v20.5.1 (/opt/homebrew/Cellar/node/20.5.1/bin/node)                            │
  │ Specs:          1 found (create-experiment.feature)                                            │
  │ Searched:       **/create-experiment.feature                                                   │
  │ Experiments:    experimentalMemoryManagement=true,experimentalModifyObstructiveThirdPartyCode… │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

overall test execution time is under 10 minutes.

**Comment by jennifer-shehane** [comment]
@rakeshnambiar Please open a new issue and share a way to reproduce.

**Comment by rakeshnambiar** [comment]
@jennifer-shehane thanks for your reply. Considering I cannot share my current script / project details, I need to find a public website match similar scenario. However, what I know so far, if I run a single tests which last more than 30 minutes, that is slowing down the execution and mostly failing due to the damn slowness of the test. At the same time if I access the application, it is not slow and everything is fine. So it is something to do with Cypress memory management. As a work-around, I just splitted this into two tess. After that, tests are running fine and it finished the execution within 20 minutes. I may need to spent sometime over the Xmas holidays to share a sample repo if at all I managed to make a sample.
